liquid-compiler/src/parser.rs

```diff
@@ -3,622 +3,411 @@
-            }
-            _ => (),
-        }
-
-        expect(&mut iter, &Token::Colon)?;
+    let literal = literal
+        .into_inner()
+        .next()
+        .expect("Get into the rule inside literal.");
 
-        // loops through the argument list after the filter name
-        while iter.peek() != None && iter.peek().unwrap() != &&Token::Pipe {
-            args.push(iter.next().unwrap().to_arg()?);
+    match literal.as_rule() {
+        Rule::StringLiteral => {
+            let literal = literal.as_str();
+            let trim_quotes = &literal[1..literal.len() - 1];
 
-            // ensure that the next token is either a Comma or a Pipe
-            match iter.peek() {
-                Some(&&Token::Comma) => {
-                    let _ = iter.next().unwrap();
-                    continue;
-                }
-                Some(&&Token::Pipe) | None => break,
-                _ => {
-                    return Err(unexpected_token_error(
-                        "`,` | `|`",
-                        Some(iter.next().unwrap()),
-                    ));
-                }
-            }
+            Scalar::new(trim_quotes.to_owned())
         }
+        Rule::IntegerLiteral => Scalar::new(
+            literal
+                .as_str()
+                .parse::<i32>()
+                .expect("Matches are parseable as integers."),
+        ),
+        Rule::FloatLiteral => Scalar::new(
+            literal
+                .as_str()
+                .parse::<f64>()
+                .expect("Matches are parseable as floats."),
+        ),
+        Rule::BooleanLiteral => Scalar::new(
+            literal
+                .as_str()
+                .parse::<bool>()
+                .expect("Matches are parseable as bools."),
+        ),
+        _ => unreachable!(),
+    }
+}
 
-        filters.push(FilterCall::new(name, args));
+fn parse_variable(variable: Pair) -> Variable {
+    if variable.as_rule() != Rule::Variable {
+        panic!("Expected variable.");
     }
 
-    Ok(FilterChain::new(entry, filters))
+    let mut indexes = variable.into_inner();
+
+    let first_identifier = indexes
+        .next()
+        .expect("A variable starts with an identifier.")
+        .as_str()
+        .to_owned();
+    let mut variable = Variable::with_literal(first_identifier);
+
+    let indexes = indexes.map(|index| match index.as_rule() {
+        Rule::Identifier => Expression::with_literal(index.as_str().to_owned()),
+        Rule::Value => parse_value(index),
+        _ => unreachable!(),
+    });
+
+    variable.extend(indexes);
+    variable
+
+    /* Why does this code appear on parse_expression() from the old parser?
+     * ```
+     * Some(&Token::Identifier(ref x)) if options.tags.contains_key(x.as_str()) => {
+     *      options.tags[x.as_str()].parse(x, &tokens[1..], options)
+     * }
+     * ```
+     * Are {{ expressions }} supposed to be treated as {% tags %} just because they have a tag name?
```

Comment:
- Issue: https://github.com/cobalt-org/liquid-rust/issues/40
- PR: https://github.com/cobalt-org/liquid-rust/pull/43
- Commit: https://github.com/cobalt-org/liquid-rust/commit/d2166407b6af0646169a7ade8deee6a9b7489246

I also dug into the Ruby code for liquid; I didn't do a good job of identifying if this behavior is in it.

My overall suspicion is that when filters were added to `assign`, the code was reused but in a way that accidentally made us also support tags for `{{}}` expressions.

We should note a breaking change for removing this.